### DISCUSSION TITLE
$http.send returns stream error

### DISCUSSION THREAD

**Discussion by LindezaGrey** [pocketbase/pocketbase#5288] [Q&A] [answered]
Hi there,

i am using $http.send to get different pages on the internet. One of the pages gives me a strange error.

When i do:

```js
 const res = $http.send({
        url: "https://www.berlin.de/",
        method: "GET",
    })
```
i receive the error:

> Something went wrong while processing your request.
   └─ Get "https://www.berlin.de/": stream error: stream ID 1; INTERNAL_ERROR; received from peer

Based on other discussion i tried to set headers or set the body to null.

Pocketbase is running in a container. If i use `curl` in the same container with 

`curl "https://www.berlin.de/"`

i receive a proper response.
My suspicion is it may have something to do with http or tls versions. I tried to analyze it with Wireshark but that didn't leave me with much additional information except that in fact the packages sent by pocketbase for the GET operation are a bit smaller than if it is done with curl. Any clue what i could change or test?

best regards

**Comment by ganigeorgiev** [comment]
Hm, I'm not sure what could be the cause of the issue here.

Under the hood `$http.send` is using the std Go [`net/http.DefaultClient`](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/net/http/client.go;l=109) and will transparently switch to HTTP2 for us. You can try to enable verbose logging with `GODEBUG=http2debug=2 ./pocketbase serve`. Based on the error message your issue is maybe related to https://github.com/golang/go/issues/51323 or https://github.com/golang/go/issues/36759 but as of now doesn't seem to have known root cause and it could be an incompatibility between the Go client and the website server implementation.

You can force disable the http2 client using `GODEBUG=http2client=0` but it depends on the website server whether it could serve its content via http1 or not.

**Reply by ganigeorgiev** [reply]
There are 2 issues here but both seems to be because of some firewall rule.

1. You need to set a `User-Agent` different from the default Go one, for example: `"User-Agent": "curl/7.85.0"`.
2. The `Content-Type` header should not be send with the request (or at least I cannot find a value that allows me to send a request to the website).

At the moment `$http.send` doesn't allow unsetting the `Content-Type` (even if it is missing it will set a default one `application/json`). This is already changed with the refactoring and we no longer set a default `Content-Type` but I don't want to push it as a minor release since it could be a breaking change. I'm not sure what to suggest for this at the moment. It doesn't work also with curl: `curl -H "Content-Type: text/html" "https://www.berlin.de/"`.

**Reply by ganigeorgiev** [reply, marked as answer]
As an alternative for this specific site you could also try sending the request as `POST` since in this case the `Content-Type` header is allowed, aka:

```js
const res = $http.send({
    url: "https://www.berlin.de/",
    method: "POST",
    headers: {
        "User-Agent":   "curl/7.85.0", // anything different from the default Go one
        "content-type": "text/plain",
    },
})
```
_(at first glance it seems to result in the same content as GET)_

Another temp option could be to call `curl` with the [`$os.cmd`](https://pocketbase.io/jsvm/functions/_os.cmd.html).

**Reply by LindezaGrey** [reply]
thanks for your fast and thorough investigation! 
I will try to debug a little more with your suggestions but i stumbled upon the issues you mentioned above...
I could also imagine that it really is some kind of "security measure" against scraping or just incompetence haha 😆 

I really like the hacky version of just using POST requests. But it seems that this does not work on all paths of the website - probably only for the root page.

I was thinking about using $os.cmd and your suggestion is the affirmation for me to go down this route - at least for now. If i find a handier solution or the main cause of this issue i will post an update.

THANKS again - keep up your good work!

**Reply by LindezaGrey** [reply]
for the sake of completeness and future reference -> i solved it with a helper function, that uses `$os.cmd`.
Here is a minimalistic implementation for the `GET` method,

```js
function axios(url) {
    const filename = "/tmp/axios.html"
    $os.cmd("curl", ...["-o", filename, url]).run()
    // return $template.loadFiles("/tmp/axios.html").render()
    // see answer below
}
```

**Reply by ganigeorgiev** [reply]
I guess it would be fine if you want to store the result in a file but I don't recommend rendering it as a template as this could have both performance and security implications (_Go templates allow running functions, accessing variables, etc. and loading an untrusted file content as raw template should be avoided_).

If you just want to return the content of the url as string you could try:

```js
function axios(url) {
    // note: make sure that "url" is a url to prevent rce
    const cmd = $os.cmd('curl', url)
    return String.fromCharCode(...cmd.output());
}
```

**Reply by LindezaGrey** [reply]
yes your version works better! thanks for the follow up
